oci/oci.go

```diff
@@ -32,35 +32,47 @@ func SetCapabilities(s *specs.Spec, caplist []string) error {
 	return nil
 }
 
-// AppendDevicePermissionsFromCgroupRules takes rules for the devices cgroup to append to the default set
-func AppendDevicePermissionsFromCgroupRules(devPermissions []specs.LinuxDeviceCgroup, rules []string) ([]specs.LinuxDeviceCgroup, error) {
+// DevicePermissionsFromCgroupRules takes rules for the devices cgroup
+func DevicePermissionsFromCgroupRules(rules []string) ([]specs.LinuxDeviceCgroup, error) {
+	var devPermissions []specs.LinuxDeviceCgroup
 	for _, deviceCgroupRule := range rules {
 		ss := deviceCgroupRuleRegex.FindAllStringSubmatch(deviceCgroupRule, -1)
 		if len(ss) == 0 || len(ss[0]) != 5 {
 			return nil, fmt.Errorf("invalid device cgroup rule format: '%s'", deviceCgroupRule)
 		}
 		matches := ss[0]
 
+		if matches[0] == "a" || matches[1] == "a" {
+			ms := deviceCgroupARuleRegex.MatchString(matches[0])
+			if !ms {
+				return nil, fmt.Errorf("although this cgroup rule is technically correct, because 'a' maps to 'a *:* rwm' regardless of what comes next, this format is partially ineffective: '%s'", deviceCgroupRule)
+			}
+
+			// TODO Is the type field needed?
+			dPermissions := specs.LinuxDeviceCgroup{
+				Allow:  true,
+				Type:   "a",
+				Access: "rwm",
+			}
+
+			devPermissions = append(devPermissions, dPermissions)
+			return devPermissions, nil
+		}
+
 		dPermissions := specs.LinuxDeviceCgroup{
 			Allow:  true,
 			Type:   matches[1],
 			Access: matches[4],
 		}
-		if matches[2] == "*" {
-			major := int64(-1)
-			dPermissions.Major = &major
-		} else {
-			major, err := strconv.ParseInt(matches[2], 10, 64)
+		if matches[2] != "*" {
+			major, err := strconv.ParseInt(matches[2], 10, 12)
```

Comment:
I'm wondering if `12 bit` is correct here. I recall the kernel has some "weird" things; e.g., using an "unsigned" integer, but allowing `-1` as "special value". In this case (just from a quick look); it looks like it's indeed using `12 bit unsigned` and `20 bit unsigned` for these; https://github.com/torvalds/linux/blob/139711f033f636cc78b6aaf7363252241b9698ef/include/linux/kdev_t.h#L7-L11

Which would mean (`-1`?) _or_ `0-4096` / `0-1048576` ?

While definitely not the "canonical source of truth", this StackOverflow question has an interesting quote as well; https://stackoverflow.com/a/14849363/1811501 (emphasis mine)

> As of Version 2.6.0 of the kernel, dev_t is a 32-bit quantity with 12 bits set aside for the major number and 20 for the minor number. ***Your code should, of course, never make any assumptions about the internal organization of device numbers; it should, instead, make use of a set of macros found in linux/kdev_t.h***

Reply:
Yeah, that does seem to be the case.